### DISCUSSION TITLE
GD-4.2.2: First time user. Tests run with many error messages

### DISCUSSION THREAD

**Discussion by SamuelAsherRivello** [MikeSchulze/gdUnit4#519] [General]
### The used GdUnit4 version

GdUnit4 4.3.2

### The used Godot version

v4.2.2-stable-mono

### Operating System

Windows 11

### Describe the bug

Hiya,

Great work on a great tool. I'm experiencing some issues. I'm using the latest public windows 11, the latest public godot, the latest version of your tool (From asset lib) that then updated itself. I am using the latest public version of rider and had to follow your docs and manually edit the csjproject. All compiles. The tests run.

Something unique in my project setup is that the "RMC Mingle" and "RMC Mini Mvcs" in my addons folder are local windows simlinks to improve my addon development workflow. I doubt that causes issues per below, but thought I'd mention it.

--

I'm a game dev with over 20 years of XP, the last decade in Unity.

I have created Mvcs architectures (https://bit.ly/mvc-architecture-for-unity-on-udemy) and Unit Testing courses (https://bit.ly/unit-testing-for-unity-course) for Unity.

I'm playing with Godot and enjoying it. I'd like to do a free YouTube video on "Unit Testing in Godot" and have tried several C#-friendly Godot unit testing pathways including yours.

I'd love some help with these bugs. 

-Sam!

Feel free to say hi - https://www.linkedin.com/in/samuelasherrivello/

Happy to have a video chat too. 

## SETUP 

Here is my godot version
![2024-06-24_13-05-47](https://github.com/MikeSchulze/gdUnit4/assets/63511769/573d9c06-25ac-4cf8-b9db-3421adc65ccf)

My default CS Project shows errors.
![2024-06-24_13-21-38](https://github.com/MikeSchulze/gdUnit4/assets/63511769/220cb96e-b369-4902-979b-3c160da381a9)

I updated to this and the errors go away. The rest of the images assume this setup
![2024-06-24_13-22-05](https://github.com/MikeSchulze/gdUnit4/assets/63511769/9c4dfe17-4279-4209-89af-499c42324c8e)

## BUGS

The tests do indeed run properly. None of the tests appear in the "GDUnit" tab on the left which I assume is unexpected.
![2024-06-24_13-06-01](https://github.com/MikeSchulze/gdUnit4/assets/63511769/8872223e-e1bf-49f5-abe6-257ff962657b)

The output is full of errors
![2024-06-24_13-06-07](https://github.com/MikeSchulze/gdUnit4/assets/63511769/4d9805e1-3eb4-4091-aad1-bec03156716a)

And I cannot edit any shortcuts
![2024-06-24_13-06-36](https://github.com/MikeSchulze/gdUnit4/assets/63511769/d5a59d68-66ea-467b-9ab4-05609dcfdbb1)

And I see some warnings
![2024-06-24_13-18-10](https://github.com/MikeSchulze/gdUnit4/assets/63511769/2a55f5fd-7b9e-4116-8fad-99f045c6e069)

### Steps to Reproduce

See images

### Minimal reproduction project

See images

**Comment by MikeSchulze** [comment]
Hi and welcome to GdUnit4.

Seems to me to be a wrong project setup, please check your C# project settings.
```
<PropertyGroup>
    <TargetFrameworks>net8.0</TargetFrameworks>
    <LangVersion>11.0</LangVersion>
    <!--Force nullable warnings, you can disable if you want-->
    <Nullable>enable</Nullable>
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
    <!--Disable warning of invalid/incompatible GodotSharp version-->
    <NoWarn>NU1605</NoWarn>
  </PropertyGroup>
  <ItemGroup>
    <!--Required for GdUnit4-->
    <PackageReference Include="gdUnit4.api" Version="4.2.*" />
  </ItemGroup>
  ```
  
  https://mikeschulze.github.io/gdUnit4/csharp_project_setup/csharp-setup/#gdunit4-c-test-setup

**Comment by SamuelAsherRivello** [comment]
Thanks!

I appreciate the help. Its improving, but my issues continue.

## SETUP

I tried your csproject as is. I deleted my csproject and solution. Added your snippet. closed/opened godot. Rebuild the project. And run the testing. Issues continue.

Then I tried a custom version from the snippet below for less warnings. I deleted my csproject and solution. Added your snippet. closed/opened godot. Rebuild the project. And run the testing. Issues continue. The following images are with my snippet here.

```
<Project Sdk="Godot.NET.Sdk/4.2.2">
  <PropertyGroup>
    <TargetFrameworks>net8.0</TargetFrameworks>
    <LangVersion>11.0</LangVersion>
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
    <!--Disable warning of invalid/incompatible GodotSharp version-->
    <NoWarn>NU1605</NoWarn>

    <!--Disable warnings. SamR preferences-->
    <!-- <Nullable>enable</Nullable> -->
    <NoWarn>CS8601</NoWarn>
    <NoWarn>CS8618</NoWarn>
    <NoWarn>CS8601</NoWarn>
    <NoWarn>CS8632</NoWarn>

  </PropertyGroup>
  <ItemGroup>
    <!--Required for GdUnit4-->
    <PackageReference Include="gdUnit4.api" Version="4.2.*" />
  </ItemGroup>
</Project>
```

## BUGS

I have one test class. Its the default from the docs. The test 'works' but does not appear in the GDUnit window in the upper left and it shows errors in the console.
![2024-06-24_13-59-23](https://github.com/MikeSchulze/gdUnit4/assets/63511769/107a2368-6a99-47ba-8061-9f831d9e8904)

And I cannot adjust any shortuts.
![2024-06-24_13-59-43](https://github.com/MikeSchulze/gdUnit4/assets/63511769/97b2741e-853a-4f9d-9957-ab849e961445)

**Comment by MikeSchulze** [comment]
hm thats strange, 
can you close Godot
delete the .godot folder
restart godot two time and rebuild the project

**Comment by SamuelAsherRivello** [comment]
Thanks!

Overall, the C#-compatibility is someone unique in the unit testing space, and the API for testing (asserts) are very nice. I'd love to use this for some upcoming videos if I can get it working. Not too urgent. But this week, I'll record with whatever testing setup I find that works best.

I tried that in my main repo. Same Results.

I'm new to Godot (< 100 hours). I'm new to addons. Perhaps I'm doing something silly and incorrect.

Below is a new fresh public repo (without the complexity of simlinks I mentioned earlier) and no gitignore so you see all the files.  I tried your suggestions in this new temp repo. Same Results. 

Again, I can right click the *.cs and run the tests and see valuable pass/fail results.  But the upper-left window seems to be broken, the output seems to be too-full of errors, and (while many non-shortcut GDUnit4 settings appear to 'work') the GDUnit4 settings for shortcuts do not appear interactive/clickable/changeable.

If you have any ideas, or find something simple I'm missing, please send a PR request with the change. I'll use it immediately. Thanks so much!

https://github.com/SamuelAsherRivello/godot-unit-test-temporary

**Comment by MikeSchulze** [comment]
Hi, I'll take a look at your repository and hopefully figure out what's going wrong here.
It would help me if you give a full list of steps to reproduce, sounds like if you are having various issues.

Kind regards
Mike

## Update
Ok, with your project, I'm able to see the errors now.
I still investigate what happens, first issue is the test itself is not found by the test discovery.

The test discover can't find your tests because of misleading test root folder.
The default folder is `test` but your tests are located under `Tests`
You need to change your test root folder to enable finding tests.
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/b2bedccb-93bb-4047-b7f7-a3fbbb968771)
But it should work also when using the context menu to run/debug tests



When you develop with C# I guess you are using a different IDE to edit your classes.
You know about the gdunit4.test.adapter?
e.g. on Rider
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/bc7cce78-bcb9-415e-911c-d754e8536450)

Checkout the [gdUnit4Net](https://github.com/MikeSchulze/gdUnit4Net) project and supported IDE's

You need to extend your project by the this lines.
```
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
    <PackageReference Include="gdUnit4.api" Version="4.2.*" />
    <PackageReference Include="gdUnit4.test.adapter" Version="1.*" />
  </ItemGroup>
```


## issues
https://github.com/MikeSchulze/gdUnit4/issues/517
https://github.com/MikeSchulze/gdUnit4/issues/518

**Comment by SamuelAsherRivello** [comment]
Cool. Here you go. If you have an solutions, please send a PR request to https://github.com/SamuelAsherRivello/godot-unit-test-temporary

### Repro Steps

## 1. Play From GdUnit Window

1. Download https://github.com/SamuelAsherRivello/godot-unit-test-temporary
1. Open the godot project
1. Press play in the "GdUnit" window. 
1. Expected Result: No errors in the "Output" window
1. Actual Result: Many errors in the "Output" window

## 2. Play From Right Click Menu

1. Download https://github.com/SamuelAsherRivello/godot-unit-test-temporary
1. Open the godot project
1. Right Click "BaseControllerTest.cs" in the FileSystemWindow. Choose "Run TestSuites"
1. Expected Result: No errors in the "Output" window
1. Actual Result: Many errors in the "Output" window

# Answers

* I use Rider as my code editor. 
* I do not know about gdunit4.test.adapter. If setting up that is a requirement to solve my bug reports above, please let me know more. If its optional, then disregard.

**Reply by SamuelAsherRivello** [reply]
I changed the "Test Lookup Folder" settings value from "tests" (default) to "" in the project settings as you mentioned. The Bug report 1 and 2 are now fixed.

Thanks for the help. Awesome plugin.

IMHO...
* When "Test Discovery" is false (default), I would assume the "Test Lookup Folder" value is irrelevant. Apparently it is indeed relevant.
* Since I assume the benefit of setting it is purely an optimization (e.g. of CPU cycles) I suggest that the "Test Lookup Folder" settings value be "" as default. Thus optional optimization is opt-in and core functionality is opt-out.

Thanks!

**Reply by MikeSchulze** [reply]
No, the test root folder is not irrelevant!
The main reason for the test root folder is to structure your project to split source and test code.
It is also used to create tests in the right folder structure (using context menu `Create Test`).

**Comment by SamuelAsherRivello** [comment]
This issue persists.

# Repro Steps

## 3. Setting Shortcuts

1. Download https://github.com/SamuelAsherRivello/godot-unit-test-temporary
1. Open the godot project
1. Click "GdUnit Settings" button from the "GdUnit" Window
1. Click the "Shortcuts" tab
1. Click any shortcut or click any shortcut THEN keyboard input after clicking
1. Expected Result: Clicking a shortcut allows the value to change.
1. Actual Result: Clicking a shortcut is unresponsive

**Reply by MikeSchulze** [reply]
I will take a look, if it is broken we need a bug issue.

https://github.com/MikeSchulze/gdUnit4/issues/522

**Comment by MikeSchulze** [comment]
> 1. Play From GdUnit Window

Do specify more details, what do you mean by `Actual Result: Many errors in the "Output" window`
Please attach the errors as script snipped, do not attach screenshots,


> 2. Play From Right Click Menu

is already covered by bug issue https://github.com/MikeSchulze/gdUnit4/issues/518

> I do not know about gdunit4.test.adapter. If setting up that is a requirement to solve my bug reports above, please let me know more. If its optional, then disregard.

The `gdunit4.test.adapter` is only need to install if you want to run/debug test in a different API e.g. Rider
Please read the [documentation](https://github.com/MikeSchulze/gdUnit4Net/blob/master/testadapter/README.md)

**Comment by SamuelAsherRivello** [comment]
1. This issue is resolved per above.


2. Great.


3. Cool. Thanks.
